freqtrade/exchange/exchange.py

```diff
@@ -1153,6 +1153,44 @@ def get_balances(self) -> dict:
             balances.pop("total", None)
             balances.pop("used", None)
 
+            if self.trading_mode == TradingMode.FUTURES:
+
+                open_orders_response: List[dict] = self._api.fetch_open_orders()
+                open_orders: dict = {}
+                for order in open_orders_response:
+                    symbol: str = order['symbol']
+                    open_orders[symbol] = order
+
+                positions: List[dict] = self._api.fetch_positions()
+                for position in positions:
+                    symbol = position['symbol']
+                    market: dict = self.markets[symbol]
+                    size: float = self._contracts_to_amount(symbol, position['contracts'])
+                    side: str = position['side']
+                    if size > 0:
+
+                        if symbol in open_orders:
+                            order = open_orders[symbol]
+                            order_amount: float = order['remaining']
+                            order_side: str = order['side']
+                            if order_side == 'buy' or order_side == 'long':
+                                order_amount = 0
+                        else:
+                            order_amount = 0
+
+                        if side == 'long' or side == 'buy':
+                            currency = market['base']
+                            free = size - order_amount
+
+                            balances[currency] = {
+                                'free': free,
+                                'used': order_amount,
+                                'total': size,
+                            }
+                            balances['free'][currency] = free
+                            balances['used'][currency] = order_amount
+                            balances['total'][currency] = size
```

Comment:
One thing is for sure - we'll not need both ways (we're actually dropping the latter way with "pop()" above, so you'll need to be careful as this will most likely crash as `balances['free']` does no longer exist as dict.
I know ccxt returns it this way - but we only look at the "currency" part (in the end it's the same content ...).

---

i'm not sure i'd do it this way though (basically messing with the currency endpoint/response).

AFAIK, we only use `get_wallets()` in one place (`wallets.py`) - so i'd keep the "position" endpoint separate (from an exchange class perspective), and instead move the logic to `wallets.py` (we'll still need the exchange class, but it'll be the usual thin wrapper around the ccxt method).

My initial idea on this was to add a 4th field (on top of free/used/total) to the Balance object - `position` - which can be positive or negative (that way we can correctly capture shorts too), and which contains the position we're currently in.

Now i'm not sure if that "position" part should contain the currency (like `DOT` in `DOT/USDT:USDT`) - or the whole pair (`DOT/USDT:USDT`) - as in theory, we could (in the future) support expiring futures, where we'd have multiple "DOT" pairs with different expiry - and the position is really a `DOT/USDT:USDT` position, not a DOT position (that's slightly different to spot markets - where you could take the DOT you bought and "run" with them to your offline wallet.

Hope this makes somewhat sense ...

Reply:
> My initial idea on this was to add a 4th field (on top of free/used/total) to the Balance object - position - which can be positive or negative (that way we can correctly capture shorts too), and which contains the position we're currently in.

I don't understand how this would be used, would the part that needs it just check if there's enough in position as well?

>  as in theory, we could (in the future) support expiring futures, where we'd have multiple "DOT" pairs with different expiry

If they have a different expiry they will have a different symbol, the expiry date is included in delivery future symbols **ex:** `BTC/USDT:USDT-220315`

> If you want i can have a look at this topic as well - as I've had a pretty clear idea about this (which will fail once i start to implement it, but we'll see ...)

If you don't feel swamped with too much other stuff, I feel like I'm running out of ways to be an asset

What pieces do you think are left before the branch "works"

Reply:
> I don't understand how this would be used, would the part that needs it just check if there's enough in position as well?

Wallets are used for 2 things
* check if we have enough
* Report it via RPC.

There is however quite some difference between "is my position there" versus "do i own enough DOT to sell in spot markets" (one being that we need to use the full symbol for positions - as you point out above).

> I feel like I'm running out of ways to be an asset

The best way to test i feel at this point is "real life tests".
This can be as easy as running a bot with a "nothing" strategy (no signal) - but setting leverage to "whatever" ...
and then use forcebuy to enter a trade (maybe with different order types) - look at the values in freqUI (are values as expected, stoploss, profit calculations, ...)
maybe try a "tight" stoploss (especially on exchange) - so it adjusts frequently, ...

I've done this so far for gate.io and okx (but will need to repeat that once the liquidation branch is ready) - but binance has efficiently blocked my futures access, so i won't be able to do any tests there.